nservicebus/upgrades/asp-saga-deduplication.md

```diff
@@ -0,0 +1,78 @@
+- concurrency is set to > 1
+
+### Symptoms
+The following are three of the symptoms that a system may exhibit as result of this bug.
+- sending more than one message when only one should be sent
+- sending no messages when one should be sent
+- creation of multiple instances of the same saga
+
+## Patch Requirements
+To deploy this fix throughout a system endpoints will need to be upgraded and saga data that has been stored by the Azure Storage persister will need to be patched.
+
+### Upgrading endpoints
+
+All endpoints using NServiceBus Azure will need to be upgraded to version 6.2.4 or higher.
+
+### Patching data
+Saga data stored in Azure will need to be patched using the `NServiceBus.AzureStoragePersistence.SagaDeduplicator` utility which can be downloaded from xxxxxxxxx.
+
+
+## Patch steps
+
+Steps:
+1. Run de-duplication utility
+2. Update NServiceBus.Azure dependency in all endpoints that use it
+3. Run de-duplication utility
+
+
+NOTE: Minimizing the time between steps one and two will minimize the work required later in the process.
+
+### Upgrade endpoint configurations
+Update references to NServiceBus.Azure to version 6.2.4 or later.
+
+### Running the de-duplication utility
+1. Download the de-duplication tool from xxxxxxxxx
+2. Run the de-duplication tool from any computer that has both an internet connection and the .NET Framework v4.5.2 installed
+3. Open a commandline and run the following command: `NServiceBus.AzureStoragePersistence.SagaDeduplicator.exe directory=<directory> sagaTypeName=<sagaTypeName> sagaProperty=<sagaProperty> operation=Download connectionString=<connectionString>`
+	- **directory**: This should be the full path of the working directory that the de-duplication tool will use. The tool will need to have read, write and delete permissions on the folder selected. If there are any spaces in the directory path this value should be enclosed in double quotes. Example: `directory="C:\my working folder\deduplication"`
+	- **sagaTypeName**:
+	- **sagaProperty**:
+	- **operation**: "Download"
+	- **connectionString**: This parameter is optional and there are two ways to provide the Azure connection string.
+		1. As a commandline parameter. Example: `connectionString=DefaultEndpointsProtocol=https;AccountName=anAzureStorageAccount;AccountKey=FuiDeGgJoHackI4h1d44LFK0ceqfkxDMO79In+ofWG42rw9kNDyQuErNmJ8WyJ7GLa+xy5NVaN21lOMZUevaQw==;BlobEndpoint=https://anAzureStorageAccount.blob.core.windows.net/;TableEndpoint=https://anAzureStorageAccount.table.core.windows.net/;QueueEndpoint=https://anAzureStorageAccount.queue.core.windows.net/;FileEndpoint=https://anAzureStorageAccount.file.core.windows.net/`
+		2. As a connectionStrings entry in the `NServiceBus.AzureStoragePersistence.SagaDeduplicator.exe.config`. Example:  
+```		
+<configuration>  
+  <connectionStrings>  
+		<add name="sagas" connectionStrings="--anAzureStorageConnectionString--"/>  
+	</connectionStrings> 
+</configuration>
+```
+4. what about conflicts?
```

Comment:
How about the following?

---

The deduplication tool will download conflicting sagas to the `directory\sagaTypeName`. Every set of conflicting sagas is downloaded to a separate directory containing files named with conflicting sagas identifiers. For example, running the tool with `directory=data` and `sagaTypeName=TwoInstanceSagaState` will result in a following structure:

```
data
└───TwoInstanceSagaState
    └───d10c2f15-06d2-1370-e0ee-781710b5d598
        └───0e36dc9a-eec0-455c-a4d3-b8b275711d15
        └───8ee2f4b2-eaf2-4d12-a87e-a5e000aaa815
```
where `0e36dc9a-eec0-455c-a4d3-b8b275711d15` and `8ee2f4b2-eaf2-4d12-a87e-a5e000aaa815` contain JSON payload of conflicting sagas with an additional property ` "$Choose_this_saga": false`. For each set of conflicting sagas, one of them should be updated accordingly and marked as the chosen with ` "$Choose_this_saga": true`.
Once it's done, the tool should be executed again with the `operation=Upload` option to upload selected and fixed sagas.